src/Interpreters/DatabaseCatalog.cpp

```diff
@@ -930,6 +939,100 @@ void DatabaseCatalog::enqueueDroppedTableCleanup(StorageID table_id, StoragePtr
         (*drop_task)->schedule();
 }
 
+void DatabaseCatalog::dequeueDroppedTableCleanup(StorageID table_id)
+{
+    String latest_metadata_dropped_path;
+    String table_metadata_path;
+    StorageID droped_table_id = table_id;
+    TablesMarkedAsDropped::iterator table;
+    {
+        std::lock_guard lock(tables_marked_dropped_mutex);
+        table = tables_marked_dropped.end();
+        time_t latest_drop_time = std::numeric_limits<time_t>::min();
+        for (auto it = tables_marked_dropped.begin(); it != tables_marked_dropped.end(); ++it)
+        {
+            if (it->table_id.uuid == table_id.uuid)
+            {
+                table = it;
+                break;
+            }
+
+            if (it->table_id.table_name == table_id.table_name &&
+                it->table_id.table_name == table_id.table_name &&
+                it->drop_time >= latest_drop_time)
+            {
+                latest_drop_time = it->drop_time;
+                table = it;
+            }
+        }
+
+        if (table == tables_marked_dropped.end())
+            throw Exception(ErrorCodes::UNKNOWN_TABLE,
+                "The drop task of table {} is in progress or has been dropped or has not metadata file, like Memory engine",
+                table_id.getNameForLogs());
+
+        latest_metadata_dropped_path = table->metadata_path;
+        droped_table_id = table->table_id;
+        table_metadata_path = getPathForMetadata(droped_table_id);
+
+        tables_marked_dropped.erase(table);
+        if (tables_marked_dropped_ids.contains(droped_table_id.uuid))
+            tables_marked_dropped_ids.erase(droped_table_id.uuid);
+
+        fs::rename(latest_metadata_dropped_path, table_metadata_path);
```

Comment:
why not rename at the end?
what if create fails?

Reply:
@xiedeyantu again, I would move rename at the end because metadata_dropped means we were planning on deleting it but still need to keep metadata until it's done.
So if a server shutdown while drop was in progress, on start we will still load the table but we know it needs to be dropped.
In your case, you move the table back to `metadata` folder, and then attaching, parsing or something fails which means failed `UNDROP` but if server restarts at this point it would be loaded as a table that was not dropped.